tools/carbondb_compress.py

```diff
@@ -0,0 +1,55 @@
+import faulthandler
+faulthandler.enable()  # will catch segfaults and write to stderr
+
+from lib.db import DB
+
+def compress_data():
+    query = """
+        INSERT INTO carbondb_energy_data_day (
+            type,
+            company,
+            machine,
+            project,
+            tags,
+            date,
+            energy_sum,
+            co2_sum,
+            carbon_intensity_avg,
+            record_count
+        )
+        SELECT
+            e.type,
+            e.company,
+            e.machine,
+            e.project,
+            e.tags,
+            DATE_TRUNC('day', TO_TIMESTAMP(e.time_stamp / 1000000)),
+            SUM(e.energy_value) AS energy_sum,
+            SUM(e.co2_value) AS co2_sum,
+            AVG(e.carbon_intensity) AS carbon_intensity_avg,
+            COUNT(*) AS record_count
+        FROM
+            carbondb_energy_data e
+        LEFT JOIN
+            carbondb_energy_data_day d ON e.machine = d.machine
+                                        AND e.project = d.project
+                                        AND DATE_TRUNC('day', TO_TIMESTAMP(e.time_stamp / 1000000)) = d.date
```

Comment:
what is `d.date` ? The table d is nowhere defined if I am not overlooking something.

Reply:
LEFT JOIN clause: This joins another table to the primary table mentioned in the FROM clause. The table being joined is carbondb_energy_data_day, aliased as d. A LEFT JOIN ensures that all records from the left table (carbondb_energy_data aliased as e) are included in the joined table, even if there is no matching record in the right table (carbondb_energy_data_day aliased as d).
ON clause: This specifies the condition used to match rows from the two tables being joined. There are three conditions here, all of which must be true for the rows to be considered a match:
e.machine = d.machine: The machine column of both tables must have the same value.
e.project = d.project: The project column of both tables must have the same value.
DATE_TRUNC('day', TO_TIMESTAMP(e.time_stamp / 1000000)) = d.date: This is a bit more complex. It involves converting a timestamp value in the e.time_stamp column to a date, and then truncating it to the day component, to match it against the d.date column. Here's what each function does:
TO_TIMESTAMP(e.time_stamp / 1000000): Since the time_stamp column is likely in microseconds (given the division by 1,000,000), this converts the time_stamp value from microseconds to a PostgreSQL timestamp.
DATE_TRUNC('day', ...): This truncates the timestamp obtained in the previous step to the start of the day (00:00:00), removing the time component.
The resulting date is then compared to the date column in the d table. If they are the same, the row from e is considered to match with the row from d.
The purpose of this join is to correlate entries in the carbondb_energy_data table with corresponding entries in the carbondb_energy_data_day table based on matching machine and project identifiers, as well as ensuring that the dates (ignoring time beyond the day) match between the two tables. This allows for combining or comparing daily aggregated data (from carbondb_energy_data_day) with more granular timestamped data (from carbondb_energy_data).

All my words in form of ChatGPT.

Reply:
Thanks for the lenghty explanation. My question was actually quite simpler. I know all the mechanis but searched for the "AS" keyword. I have never seen an alias without it. But now know that `LEFT JOIN table as t` = `LEFT JOIN table t`